src/rust/ensogl/lib/shape-utils/src/dynamic_color.rs

```diff
@@ -0,0 +1,277 @@
+        self.color_source.replace(source);
+    }
+
+    /// Return the path where we look for variant colors in the theme.
+    fn variant_path(path:Path, extension:String) -> Path {
+        let segments_rev = path.rev_segments;
+        let mut segments = segments_rev.into_iter().rev().collect_vec();
+        segments.pop();
+        segments.push(" variant ".to_string());
+        segments.push(extension);
+        Path::from_segments(segments)
+    }
+
+    fn try_get_color_variant_from_theme(&self, id:&str) -> Option<color::Rgba> {
+        if let Source::Theme{ path } = self.color_source.borrow().clone() {
+            let path  = Self::variant_path(path,id.to_string());
+            let color = self.styles.get(path).color()?;
+            Some(color::Rgba::from(color))
+        } else {
+            None
+        }
+    }
+
+    fn get_base_color(&self) -> color::Rgba {
+        match self.color_source.borrow().clone() {
+            Source::Static{color} => color,
+            Source::Theme{path}   => self.styles.get_color(path).into(),
+        }
+    }
+
+    /// Create a dimmed version of the given color value. The exact values to be used for dimming
+    /// are derived from the theme.
+    fn make_dimmed_color(&self, color:color::Rgba) -> color::Rgba {
+        let color : color::Lcha    = color.into();
+        let color_lightness_factor = theme::vars::graph_editor::colors::default::dimming::lightness_factor;
+        let color_chroma_factor    = theme::vars::graph_editor::colors::default::dimming::chroma_factor;
+        let color_lightness_factor = self.styles.get_number_or(color_lightness_factor,0.0);
+        let color_chroma_factor    = self.styles.get_number_or(color_chroma_factor,0.0);
+        let lightness              = color.lightness * color_lightness_factor;
+        let chroma                 = color.chroma * color_chroma_factor;
+        let color                  = color::Lcha::new(lightness,chroma,color.hue,color.alpha);
+        color.into()
+    }
+
+    /// Return the modified version of the base color.
+    fn get_parametrized_color(&self, dimmnes:f32, alpha:f32) -> color::Rgba {
+        let base_color = self.get_base_color();
+        // Check whether there is a version defined in the theme, otherwise we create our own.
+        let dimmed_color = match self.try_get_color_variant_from_theme(THEME_KEY_DIMMED) {
+            None        => self.make_dimmed_color(base_color),
+            Some(color) => color,
+        };
+        let mut color = mix(base_color,dimmed_color,dimmnes);
+        color.data.alpha *= alpha;
+        color.into()
+    }
+}
+
+
+
+// =====================
+// === Dynamic Color ===
+// =====================
+
+/// The `DynamicColor` provides color information through an FRP api. It allows for dynamic color
+/// transitions between different states (e.g, dim or not dim) that are emitted like an animation.
+#[derive(Clone,CloneRef,Debug)]
+pub struct DynamicColor {
```

Comment:
As we agreed on Discord, lets implement `Animation<Lcha>` + `style.dimmed(color)` + `FrpStyleWatch` + some small wrapper around it instead. This is a big change, so after it I'll do re-review